ipaclient/plugins/cert.py

```diff
@@ -32,12 +36,96 @@
 
 @register(override=True, no_fail=True)
 class cert_request(MethodOverride):
+    takes_options = (
+        Str(
+            'database?',
+            label=_('Path to NSS database'),
+            doc=_('Path to NSS database to use for private key'),
+        ),
+        Str(
+            'private_key?',
+            label=_('Path to private key file'),
+            doc=_('Path to PEM file containing a private key'),
+        ),
+        Flag(
+            'autogenerate',
+            label=_('Automatically generate CSR for request using LDAP data'),
+        )
+    )
+
     def get_args(self):
         for arg in super(cert_request, self).get_args():
             if arg.name == 'csr':
-                arg = arg.clone_retype(arg.name, File)
+                arg = arg.clone_retype(arg.name, File, required=False)
             yield arg
 
+    def forward(self, *keys, **options):
+        autogenerate = options.pop('autogenerate', False)
+        database = options.pop('database', None)
+        private_key = options.pop('private_key', None)
+
+        if autogenerate:
+            if database:
+                helper = u'certutil'
+                helper_args = ['-d', database]
+            elif private_key:
+                helper = u'openssl'
+                helper_args = [private_key]
+            else:
+                raise errors.InvocationError(
+                    format="One of 'database' or 'private_key' is required")
+
+            scriptfile = tempfile.NamedTemporaryFile(delete=False)
+            scriptfile.close()
+            csrfile = tempfile.NamedTemporaryFile(delete=False)
+            csrfile.close()
+            csrfilename = csrfile.name
+
+            # profile_id is optional for cert_request, but not for
+            # cert_get_requestdata, so pass the default explicitly when
+            # necessary
+            profile_id = options.get('profile_id')
+            if profile_id is None:
+                profile_id = self.get_default_of('profile_id')
```

Comment:
I know I suggested this to you, but I didn't realize it would require hard-coding the default profile name into `profile_id` in server-side `cert_request`. I would rather not do that and instead use `dogtag.DEFAULT_PROFILE` as the default value of `profile_id` in `cert_get_requestdata`. I'm aware that this goes against the idea of having the server decide what is the default profile, but `cert_get_requestdata` will be moved to the server, so this is only temporary.